File: frameworks/desktop/macos-appkit/SocialMediaFeed.swift
import Cocoa

class SocialMediaFeed: NSViewController {
    
    // MARK: - Properties
    private var posts: [Post] = []
    private var allPosts: [Post] = []
    private var currentFilter = "All"
    private var unreadNotifications = 3
    
    // MARK: - UI Elements
    @IBOutlet weak var filterComboBox: NSComboBox!
    @IBOutlet weak var notificationsButton: NSButton!
    @IBOutlet weak var trendingButton: NSButton!
    @IBOutlet weak var createPostTextField: NSTextField!
    @IBOutlet weak var createPostButton: NSButton!
    @IBOutlet weak var postsScrollView: NSScrollView!
    @IBOutlet weak var postsStackView: NSStackView!
    @IBOutlet weak var statusLabel: NSTextField!
    @IBOutlet weak var postCountLabel: NSTextField!
    
    // MARK: - Lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI()
        loadSamplePosts()
        updatePostsDisplay()
        updateNotificationBadge()
    }
    
    // MARK: - Setup
    private func setupUI() {
        // Configure combo box
        filterComboBox.addItems(withObjectValues: ["All", "Following", "Trending"])
        filterComboBox.selectItem(at: 0)
        filterComboBox.target = self
        filterComboBox.action = #selector(filterChanged)
        
        // Configure buttons
        notificationsButton.target = self
        notificationsButton.action = #selector(notificationsButtonTapped)
        
        trendingButton.target = self
        trendingButton.action = #selector(trendingButtonTapped)
        
        createPostButton.target = self
        createPostButton.action = #selector(createPostButtonTapped)
        
        // Configure text field
        createPostTextField.target = self
        createPostTextField.action = #selector(createPostTextFieldAction)
        
        // Configure stack view
        postsStackView.orientation = .vertical
        postsStackView.spacing = 16
        postsStackView.alignment = .fill
        postsStackView.distribution = .fill
        
        // Setup menu
        setupMenu()
    }
    
    private func setupMenu() {
        // File menu
        let fileMenu = NSMenu(title: "File")
        fileMenu.addItem(NSMenuItem(title: "New Post", action: #selector(createPostButtonTapped), keyEquivalent: "n"))
        fileMenu.addItem(NSMenuItem(title: "Export Feed", action: #selector(exportFeed), keyEquivalent: "e"))
        
        // Edit menu
        let editMenu = NSMenu(title: "Edit")
        editMenu.addItem(NSMenuItem(title: "Copy Post", action: #selector(copyPost), keyEquivalent: "c"))
        editMenu.addItem(NSMenuItem(title: "Paste Post", action: #selector(pastePost), keyEquivalent: "v"))
        
        // View menu
        let viewMenu = NSMenu(title: "View")
        viewMenu.addItem(NSMenuItem(title: "Refresh Feed", action: #selector(refreshFeed), keyEquivalent: "r"))
        viewMenu.addItem(NSMenuItem(title: "Show Notifications", action: #selector(notificationsButtonTapped), keyEquivalent: ""))
        viewMenu.addItem(NSMenuItem(title: "Show Trending", action: #selector(trendingButtonTapped), keyEquivalent: ""))
    }
    
    // MARK: - Data Loading
    private func loadSamplePosts() {
        allPosts = [
            Post(
                id: "1",
                user: User(id: "user1", name: "John Doe", username: "@johndoe", avatar: "👨‍💻", verified: true),
                content: "Just finished building an amazing React Native app! The development process was challenging but incredibly rewarding. #ReactNative #MobileDev #Tech",
                image: "📱",
                location: "San Francisco, CA",
                timestamp: "2h ago",
                likes: 42,
                comments: 8,
                shares: 3,
                isLiked: false,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "2",
                user: User(id: "user2", name: "Sarah Wilson", username: "@sarahw", avatar: "👩‍🎨", verified: false),
                content: "Beautiful sunset from my office window today. Sometimes you need to stop and appreciate the little moments. 🌅",
                image: "🌅",
                location: "New York, NY",
                timestamp: "4h ago",
                likes: 28,
                comments: 5,
                shares: 1,
                isLiked: true,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "3",
                user: User(id: "user3", name: "Mike Chen", username: "@mikechen", avatar: "👨‍💼", verified: true),
                content: "Excited to announce our new product launch! After months of hard work, we're finally ready to share it with the world. #Startup #Innovation",
                image: "🚀",
                location: "Austin, TX",
                timestamp: "6h ago",
                likes: 156,
                comments: 23,
                shares: 12,
                isLiked: false,
                isBookmarked: true,
                privacy: "public"
            ),
            Post(
                id: "4",
                user: User(id: "user4", name: "Emma Davis", username: "@emmad", avatar: "👩‍🔬", verified: false),
                content: "Just read an amazing article about the future of AI in healthcare. The possibilities are endless! What are your thoughts on this?",
                image: "🤖",
                location: "Boston, MA",
                timestamp: "8h ago",
                likes: 73,
                comments: 15,
                shares: 7,
                isLiked: true,
                isBookmarked: false,
                privacy: "public"
            ),
            Post(
                id: "5",
                user: User(id: "user5", name: "Alex Rodriguez", username: "@alexr", avatar: "👨‍🎓", verified: false),
                content: "Coffee and code - the perfect combination for a productive day! ☕️💻",
                image: "☕️",
                location: "Seattle, WA",
                timestamp: "10h ago",
                likes: 34,
                comments: 6,
                shares: 2,
                isLiked: false,
                isBookmarked: false,
                privacy: "public"
            )
        ]
        
        posts = allPosts
    }
    
    // MARK: - UI Updates
    private func updatePostsDisplay() {
        // Clear existing posts
        postsStackView.arrangedSubviews.forEach { $0.removeFromSuperview() }
        
        for post in posts {
            let postView = createPostView(for: post)
            postsStackView.addArrangedSubview(postView)
        }
        
        postCountLabel.stringValue = "\(posts.count) posts"
    }
    
    private func createPostView(for post: Post) -> NSView {
        let containerView = NSView()
        containerView.wantsLayer = true
        containerView.layer?.backgroundColor = NSColor.controlBackgroundColor.cgColor
        containerView.layer?.cornerRadius = 8
        containerView.layer?.borderWidth = 1
        containerView.layer?.borderColor = NSColor.separatorColor.cgColor
        
        let stackView = NSStackView()
        stackView.orientation = .vertical
        stackView.spacing = 12
        stackView.alignment = .fill
        stackView.distribution = .fill
        
        // Header with user info
        let headerStack = NSStackView()
        headerStack.orientation = .horizontal
        headerStack.spacing = 12
        headerStack.alignment = .centerY
        
        let avatarLabel = NSTextField(labelWithString: post.user.avatar)
        avatarLabel.font = NSFont.systemFont(ofSize: 24)
        avatarLabel.isEditable = false
        avatarLabel.isBordered = false
        avatarLabel.backgroundColor = NSColor.clear
        
        let userInfoStack = NSStackView()
        userInfoStack.orientation = .vertical
        userInfoStack.spacing = 2
        userInfoStack.alignment = .leading
        
        let nameStack = NSStackView()
        nameStack.orientation = .horizontal
        nameStack.spacing = 4
        nameStack.alignment = .centerY
        
        let nameLabel = NSTextField(labelWithString: post.user.name)
        nameLabel.font = NSFont.systemFont(ofSize: 16, weight: .bold)
        nameLabel.isEditable = false
        nameLabel.isBordered = false
        nameLabel.backgroundColor = NSColor.clear
        
        let verifiedLabel = NSTextField(labelWithString: post.user.verified ? "✓" : "")
        verifiedLabel.font = NSFont.systemFont(ofSize: 14)
        verifiedLabel.textColor = NSColor.systemBlue
        verifiedLabel.isEditable = false
        verifiedLabel.isBordered = false
        verifiedLabel.backgroundColor = NSColor.clear
        
        let usernameLabel = NSTextField(labelWithString: post.user.username)
        usernameLabel.font = NSFont.systemFont(ofSize: 14)
        usernameLabel.textColor = NSColor.secondaryLabelColor
        usernameLabel.isEditable = false
        usernameLabel.isBordered = false
        usernameLabel.backgroundColor = NSColor.clear
        
        let timestampLabel = NSTextField(labelWithString: post.timestamp)
        timestampLabel.font = NSFont.systemFont(ofSize: 14)
        timestampLabel.textColor = NSColor.secondaryLabelColor
        timestampLabel.isEditable = false
        timestampLabel.isBordered = false
        timestampLabel.backgroundColor = NSColor.clear
        
        nameStack.addArrangedSubview(nameLabel)
        nameStack.addArrangedSubview(verifiedLabel)
        
        userInfoStack.addArrangedSubview(nameStack)
        userInfoStack.addArrangedSubview(usernameLabel)
        userInfoStack.addArrangedSubview(timestampLabel)
        
        headerStack.addArrangedSubview(avatarLabel)
        headerStack.addArrangedSubview(userInfoStack)
        headerStack.addArrangedSubview(NSView()) // Spacer
        
        // Content
        let contentLabel = NSTextField(labelWithString: post.content)
        contentLabel.font = NSFont.systemFont(ofSize: 16)
        contentLabel.isEditable = false
        contentLabel.isBordered = false
        contentLabel.backgroundColor = NSColor.clear
        contentLabel.maximumNumberOfLines = 0
        contentLabel.lineBreakMode = .byWordWrapping
        
        // Image (if present)
        var imageLabel: NSTextField?
        if !post.image.isEmpty {
            imageLabel = NSTextField(labelWithString: post.image)
            imageLabel?.font = NSFont.systemFont(ofSize: 48)
            imageLabel?.isEditable = false
            imageLabel?.isBordered = false
            imageLabel?.backgroundColor = NSColor.clear
            imageLabel?.alignment = .center
        }
        
        // Location
        var locationLabel: NSTextField?
        if !post.location.isEmpty {
            locationLabel = NSTextField(labelWithString: "📍 \(post.location)")
            locationLabel?.font = NSFont.systemFont(ofSize: 14)
            locationLabel?.textColor = NSColor.secondaryLabelColor
            locationLabel?.isEditable = false
            locationLabel?.isBordered = false
            locationLabel?.backgroundColor = NSColor.clear
        }
        
        // Actions
        let actionsStack = NSStackView()
        actionsStack.orientation = .horizontal
        actionsStack.spacing = 16
        actionsStack.alignment = .centerY
        actionsStack.distribution = .fillEqually
        
        let likeButton = NSButton(title: "❤️ \(post.likes)", target: self, action: #selector(likeButtonTapped))
        likeButton.tag = posts.firstIndex(where: { $0.id == post.id }) ?? 0
        likeButton.bezelStyle = .roundRect
        likeButton.font = NSFont.systemFont(ofSize: 14)
        if post.isLiked {
            likeButton.layer?.backgroundColor = NSColor.systemPink.cgColor
        }
        
        let commentButton = NSButton(title: "💬 \(post.comments)", target: self, action: #selector(commentButtonTapped))
        commentButton.tag = posts.firstIndex(where: { $0.id == post.id }) ?? 0
        commentButton.bezelStyle = .roundRect
        commentButton.font = NSFont.systemFont(ofSize: 14)
        
        let shareButton = NSButton(title: "🔄 \(post.shares)", target: self, action: #selector(shareButtonTapped))
        shareButton.tag = posts.firstIndex(where: { $0.id == post.id }) ?? 0
        shareButton.bezelStyle = .roundRect
        shareButton.font = NSFont.systemFont(ofSize: 14)
        
        let bookmarkButton = NSButton(title: post.isBookmarked ? "🔖" : "📌", target: self, action: #selector(bookmarkButtonTapped))
        bookmarkButton.tag = posts.firstIndex(where: { $0.id == post.id }) ?? 0
        bookmarkButton.bezelStyle = .roundRect
        bookmarkButton.font = NSFont.systemFont(ofSize: 14)
        if post.isBookmarked {
            bookmarkButton.layer?.backgroundColor = NSColor.systemBlue.cgColor
        }
        
        actionsStack.addArrangedSubview(likeButton)
        actionsStack.addArrangedSubview(commentButton)
        actionsStack.addArrangedSubview(shareButton)
        actionsStack.addArrangedSubview(bookmarkButton)
        
        // Add all elements to main stack
        stackView.addArrangedSubview(headerStack)
        stackView.addArrangedSubview(contentLabel)
        
        if let imageLabel = imageLabel {
            stackView.addArrangedSubview(imageLabel)
        }
        
        if let locationLabel = locationLabel {
            stackView.addArrangedSubview(locationLabel)
        }
        
        stackView.addArrangedSubview(actionsStack)
        
        // Add stack view to container
        containerView.addSubview(stackView)
        stackView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            stackView.leadingAnchor.constraint(equalTo: containerView.leadingAnchor, constant: 16),
            stackView.trailingAnchor.constraint(equalTo: containerView.trailingAnchor, constant: -16),
            stackView.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 16),
            stackView.bottomAnchor.constraint(equalTo: containerView.bottomAnchor, constant: -16)
        ])
        
        return containerView
    }
    
    private func updateNotificationBadge() {
        if unreadNotifications > 0 {
            notificationsButton.layer?.backgroundColor = NSColor.systemRed.cgColor
            notificationsButton.layer?.cornerRadius = 4
        } else {
            notificationsButton.layer?.backgroundColor = NSColor.clear.cgColor
        }
    }
    
    // MARK: - Actions
    @objc private func filterChanged() {
        currentFilter = filterComboBox.stringValue
        filterPosts()
    }
    
    private func filterPosts() {
        // In a real implementation, this would filter posts based on the selected filter
        posts = allPosts
        updatePostsDisplay()
    }
    
    @objc private func notificationsButtonTapped() {
        let alert = NSAlert()
        alert.messageText = "Notifications"
        alert.informativeText = "You have \(unreadNotifications) unread notifications"
        alert.runModal()
    }
    
    @objc private func trendingButtonTapped() {
        let alert = NSAlert()
        alert.messageText = "Trending"
        alert.informativeText = "Top trending topics:\n#ReactNative\n#MobileDev\n#Tech\n#Startup\n#Innovation"
        alert.runModal()
    }
    
    @objc private func createPostTextFieldAction() {
        createPostButton.isEnabled = !createPostTextField.stringValue.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
    
    @objc private func createPostButtonTapped() {
        let content = createPostTextField.stringValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !content.isEmpty else { return }
        
        let newPost = Post(
            id: UUID().uuidString,
            user: User(id: "current-user", name: "You", username: "@you", avatar: "👤", verified: false),
            content: content,
            image: "",
            location: "",
            timestamp: "now",
            likes: 0,
            comments: 0,
            shares: 0,
            isLiked: false,
            isBookmarked: false,
            privacy: "public"
        )
        
        posts.insert(newPost, at: 0)
        allPosts.insert(newPost, at: 0)
        createPostTextField.stringValue = ""
        createPostButton.isEnabled = false
        updatePostsDisplay()
    }
    
    @objc private func likeButtonTapped(_ sender: NSButton) {
        let postIndex = sender.tag
        guard postIndex < posts.count else { return }
        
        posts[postIndex].isLiked.toggle()
        posts[postIndex].likes += posts[postIndex].isLiked ? 1 : -1
        
        // Update in allPosts as well
        if let allPostsIndex = allPosts.firstIndex(where: { $0.id == posts[postIndex].id }) {
            allPosts[allPostsIndex].isLiked = posts[postIndex].isLiked
            allPosts[allPostsIndex].likes = posts[postIndex].likes
        }
        
        updatePostsDisplay()
    }
    
    @objc private func commentButtonTapped(_ sender: NSButton) {
        let postIndex = sender.tag
        guard postIndex < posts.count else { return }
        
        let post = posts[postIndex]
        let alert = NSAlert()
        alert.messageText = "Comments"
        alert.informativeText = "Comments for: \(post.user.name)\n\n\(post.content)"
        alert.runModal()
    }
    
    @objc private func shareButtonTapped(_ sender: NSButton) {
        let postIndex = sender.tag
        guard postIndex < posts.count else { return }
        
        let post = posts[postIndex]
        let alert = NSAlert()
        alert.messageText = "Share"
        alert.informativeText = "Sharing: \(post.title)"
        alert.runModal()
    }
    
    @objc private func bookmarkButtonTapped(_ sender: NSButton) {
        let postIndex = sender.tag
        guard postIndex < posts.count else { return }
        
        posts[postIndex].isBookmarked.toggle()
        
        // Update in allPosts as well
        if let allPostsIndex = allPosts.firstIndex(where: { $0.id == posts[postIndex].id }) {
            allPosts[allPostsIndex].isBookmarked = posts[postIndex].isBookmarked
        }
        
        updatePostsDisplay()
    }
    
    // MARK: - Menu Actions
    @objc private func exportFeed() {
        let alert = NSAlert()
        alert.messageText = "Export Feed"
        alert.informativeText = "Feed export functionality would be implemented here."
        alert.runModal()
    }
    
    @objc private func copyPost() {
        let selectedRow = postsStackView.arrangedSubviews.count - 1 // Last post for demo
        if selectedRow >= 0 && selectedRow < posts.count {
            let post = posts[selectedRow]
            let pasteboard = NSPasteboard.general
            pasteboard.clearContents()
            pasteboard.setString(post.content, forType: .string)
        }
    }
    
    @objc private func pastePost() {
        let pasteboard = NSPasteboard.general
        if let content = pasteboard.string(forType: .string) {
            createPostTextField.stringValue = content
            createPostButton.isEnabled = !content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
    }
    
    @objc private func refreshFeed() {
        // Simulate refresh
        statusLabel.stringValue = "Feed refreshed"
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            self.statusLabel.stringValue = "Ready"
        }
    }
}

// MARK: - User Model
struct User {
    let id: String
    let name: String
    let username: String
    let avatar: String
    let verified: Bool
}

// MARK: - Post Model
struct Post {
    let id: String
    let user: User
    let content: String
    let image: String
    let location: String
    let timestamp: String
    var likes: Int
    let comments: Int
    let shares: Int
    var isLiked: Bool
    var isBookmarked: Bool
    let privacy: String
    
    var title: String {
        return content.count > 50 ? String(content.prefix(50)) + "..." : content
    }
}

File: scenarios/advanced/desktop/linux-gtk/SocialMediaFeed.py
#!/usr/bin/env python3

import gi
gi.require_version('Gtk', '3.0')
from gi.repository import Gtk, Gdk, Pango
from datetime import datetime

class SocialMediaFeed(Gtk.Window):
    def __init__(self):
        Gtk.Window.__init__(self, title="Social Media Feed")
        self.set_default_size(1000, 700)
        self.set_border_width(0)
        
        # Data
        self.posts = []
        self.all_posts = []
        self.current_filter = "All"
        self.unread_notifications = 3
        
        # Load sample data
        self.load_sample_posts()
        
        # Create UI
        self.create_ui()
    
    def create_ui(self):
        # Main container
        main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=0)
        self.add(main_box)
        
        # Header
        header_box = self.create_header()
        main_box.pack_start(header_box, False, False, 0)
        
        # Create post
        create_post_box = self.create_post_box()
        main_box.pack_start(create_post_box, False, False, 0)
        
        # Posts feed
        self.create_posts_feed(main_box)
        
        # Status bar
        status_box = self.create_status_bar()
        main_box.pack_start(status_box, False, False, 0)
    
    def create_header(self):
        header_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        header_box.set_margin_bottom(16)
        
        # Title
        title_label = Gtk.Label()
        title_label.set_markup("<span size='xx-large' weight='bold' foreground='white'>Social Media Feed</span>")
        header_box.pack_start(title_label, True, True, 0)
        
        # Controls
        controls_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        
        # Filter combo
        self.filter_combo = Gtk.ComboBoxText()
        self.filter_combo.append_text("All")
        self.filter_combo.append_text("Following")
        self.filter_combo.append_text("Trending")
        self.filter_combo.set_active(0)
        self.filter_combo.connect("changed", self.on_filter_changed)
        controls_box.pack_start(self.filter_combo, False, False, 0)
        
        # Notifications button
        self.notifications_btn = Gtk.Button(label="🔔")
        self.notifications_btn.set_tooltip_text("Notifications")
        self.notifications_btn.connect("clicked", self.on_notifications_clicked)
        controls_box.pack_start(self.notifications_btn, False, False, 0)
        
        # Trending button
        trending_btn = Gtk.Button(label="📈")
        trending_btn.set_tooltip_text("Trending")
        trending_btn.connect("clicked", self.on_trending_clicked)
        controls_box.pack_start(trending_btn, False, False, 0)
        
        header_box.pack_start(controls_box, False, False, 0)
        
        # Style header
        header_box.override_background_color(Gtk.StateFlags.NORMAL, Gdk.RGBA(0.2, 0.4, 0.8, 1.0))
        
        return header_box
    
    def create_post_box(self):
        create_post_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=16)
        create_post_box.set_margin_start(24)
        create_post_box.set_margin-end(24)
        create_post_box.set_margin-top(16)
        create_post_box.set_margin-bottom(16)
        
        # Avatar
        avatar_label = Gtk.Label()
        avatar_label.set_markup("<span size='xx-large'>👤</span>")
        create_post_box.pack_start(avatar_label, False, False, 0)
        
        # Text entry and button
        entry_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        
        self.create_post_entry = Gtk.Entry()
        self.create_post_entry.set_placeholder_text("What's on your mind?")
        self.create_post_entry.connect("activate", self.on_create_post)
        entry_box.pack_start(self.create_post_entry, True, True, 0)
        
        self.create_post_btn = Gtk.Button(label="Post")
        self.create_post_btn.connect("clicked", self.on_create_post)
        entry_box.pack_start(self.create_post_btn, False, False, 0)
        
        create_post_box.pack_start(entry_box, True, True, 0)
        
        # Style create post box
        create_post_box.override_background_color(Gtk.StateFlags.NORMAL, Gdk.RGBA(0.95, 0.95, 0.95, 1.0))
        
        return create_post_box
    
    def create_posts_feed(self, parent):
        # Scrolled window for posts
        scrolled = Gtk.ScrolledWindow()
        scrolled.set_hexpand(True)
        scrolled.set_vexpand(True)
        
        # Posts container
        self.posts_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=16)
        self.posts_box.set_margin-start(24)
        self.posts_box.set_margin-end(24)
        
        scrolled.add(self.posts_box)
        parent.pack_start(scrolled, True, True, 0)
        
        # Update posts display
        self.update_posts_display()
    
    def create_status_bar(self):
        status_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=8)
        status_box.set_margin-start(8)
        status_box.set_margin-end(8)
        status_box.set_margin-top(4)
        status_box.set-margin-bottom(4)
        
        self.status_label = Gtk.Label("Ready")
        status_box.pack_start(self.status_label, True, True, 0)
        
        self.post_count_label = Gtk.Label(f"{len(self.posts)} posts")
        status_box.pack_start(self.post_count_label, False, False, 0)
        
        # Style status bar
        status_box.override_background_color(Gtk.StateFlags.NORMAL, Gdk.RGBA(0.95, 0.95, 0.95, 1.0))
        
        return status_box
    
    def load_sample_posts(self):
        self.all_posts = [
            {
                'id': '1',
                'user': {
                    'id': 'user1',
                    'name': 'John Doe',
                    'username': '@johndoe',
                    'avatar': '👨‍💻',
                    'verified': True
                },
                'content': 'Just finished building an amazing React Native app! The development process was challenging but incredibly rewarding. #ReactNative #MobileDev #Tech',
                'image': '📱',
                'location': 'San Francisco, CA',
                'timestamp': '2h ago',
                'likes': 42,
                'comments': 8,
                'shares': 3,
                'is_liked': False,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '2',
                'user': {
                    'id': 'user2',
                    'name': 'Sarah Wilson',
                    'username': '@sarahw',
                    'avatar': '👩‍🎨',
                    'verified': False
                },
                'content': 'Beautiful sunset from my office window today. Sometimes you need to stop and appreciate the little moments. 🌅',
                'image': '🌅',
                'location': 'New York, NY',
                'timestamp': '4h ago',
                'likes': 28,
                'comments': 5,
                'shares': 1,
                'is_liked': True,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '3',
                'user': {
                    'id': 'user3',
                    'name': 'Mike Chen',
                    'username': '@mikechen',
                    'avatar': '👨‍💼',
                    'verified': True
                },
                'content': 'Excited to announce our new product launch! After months of hard work, we\'re finally ready to share it with the world. #Startup #Innovation',
                'image': '🚀',
                'location': 'Austin, TX',
                'timestamp': '6h ago',
                'likes': 156,
                'comments': 23,
                'shares': 12,
                'is_liked': False,
                'is_bookmarked': True,
                'privacy': 'public'
            },
            {
                'id': '4',
                'user': {
                    'id': 'user4',
                    'name': 'Emma Davis',
                    'username': '@emmad',
                    'avatar': '👩‍🔬',
                    'verified': False
                },
                'content': 'Just read an amazing article about the future of AI in healthcare. The possibilities are endless! What are your thoughts on this?',
                'image': '🤖',
                'location': 'Boston, MA',
                'timestamp': '8h ago',
                'likes': 73,
                'comments': 15,
                'shares': 7,
                'is_liked': True,
                'is_bookmarked': False,
                'privacy': 'public'
            },
            {
                'id': '5',
                'user': {
                    'id': 'user5',
                    'name': 'Alex Rodriguez',
                    'username': '@alexr',
                    'avatar': '👨‍🎓',
                    'verified': False
                },
                'content': 'Coffee and code - the perfect combination for a productive day! ☕️💻',
                'image': '☕️',
                'location': 'Seattle, WA',
                'timestamp': '10h ago',
                'likes': 34,
                'comments': 6,
                'shares': 2,
                'is_liked': False,
                'is_bookmarked': False,
                'privacy': 'public'
            }
        ]
        
        self.posts = self.all_posts.copy()
    
    def update_posts_display(self):
        # Clear existing posts
        for child in self.posts_box.get_children():
            self.posts_box.remove(child)
        
        # Add posts
        for post in self.posts:
            post_widget = self.create_post_widget(post)
            self.posts_box.pack_start(post_widget, False, False, 0)
        
        self.posts_box.show_all()
        self.post_count_label.set_text(f"{len(self.posts)} posts")
    
    def create_post_widget(self, post):
        # Main container
        post_frame = Gtk.Frame()
        post_frame.set_margin_bottom(16)
        
        main_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=12)
        main_box.set_margin_start(16)
        main_box.set_margin-end(16)
        main_box.set_margin-top(16)
        main_box.set_margin-bottom(16)
        
        # Header
        header_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=12)
        
        # Avatar
        avatar_label = Gtk.Label()
        avatar_label.set_markup(f"<span size='xx-large'>{post['user']['avatar']}</span>")
        header_box.pack_start(avatar_label, False, False, 0)
        
        # User info
        user_info_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL, spacing=2)
        
        # Name and verification
        name_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=4)
        
        name_label = Gtk.Label()
        name_label.set_markup(f"<span weight='bold'>{post['user']['name']}</span>")
        name_box.pack_start(name_label, False, False, 0)
        
        if post['user']['verified']:
            verified_label = Gtk.Label()
            verified_label.set_markup("<span color='blue'>✓</span>")
            name_box.pack_start(verified_label, False, False, 0)
        
        user_info_box.pack_start(name_box, False, False, 0)
        
        # Username and timestamp
        username_label = Gtk.Label()
        username_label.set_markup(f"<span color='gray'>{post['user']['username']} • {post['timestamp']}</span>")
        user_info_box.pack_start(username_label, False, False, 0)
        
        header_box.pack_start(user_info_box, False, False, 0)
        header_box.pack_start(Gtk.Label(), True, True, 0)  # Spacer
        
        main_box.pack_start(header_box, False, False, 0)
        
        # Content
        content_label = Gtk.Label()
        content_label.set_markup(f"<span>{post['content']}</span>")
        content_label.set_halign(Gtk.Align.START)
        content_label.set_line_wrap(True)
        main_box.pack_start(content_label, False, False, 0)
        
        # Image
        if post['image']:
            image_label = Gtk.Label()
            image_label.set_markup(f"<span size='xx-large'>{post['image']}</span>")
            image_label.set_halign(Gtk.Align.CENTER)
            main_box.pack_start(image_label, False, False, 0)
        
        # Location
        if post['location']:
            location_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=4)
            
            location_icon = Gtk.Label()
            location_icon.set_markup("<span color='gray'>📍</span>")
            location_box.pack_start(location_icon, False, False, 0)
            
            location_label = Gtk.Label()
            location_label.set_markup(f"<span color='gray'>{post['location']}</span>")
            location_box.pack_start(location_label, False, False, 0)
            
            main_box.pack_start(location_box, False, False, 0)
        
        # Actions
        actions_box = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=16)
        
        # Like button
        like_btn = Gtk.Button()
        like_icon = "❤️" if post['is_liked'] else "🤍"
        like_btn.set_label(f"{like_icon} {post['likes']}")
        like_btn.connect("clicked", self.on_like_clicked, post)
        actions_box.pack_start(like_btn, False, False, 0)
        
        # Comment button
        comment_btn = Gtk.Button()
        comment_btn.set_label(f"💬 {post['comments']}")
        comment_btn.connect("clicked", self.on_comment_clicked, post)
        actions_box.pack_start(comment_btn, False, False, 0)
        
        # Share button
        share_btn = Gtk.Button()
        share_btn.set_label(f"🔄 {post['shares']}")
        share_btn.connect("clicked", self.on_share_clicked, post)
        actions_box.pack_start(share_btn, False, False, 0)
        
        # Bookmark button
        bookmark_btn = Gtk.Button()
        bookmark_icon = "🔖" if post['is_bookmarked'] else "📌"
        bookmark_btn.set_label(bookmark_icon)
        bookmark_btn.connect("clicked", self.on_bookmark_clicked, post)
        actions_box.pack_start(bookmark_btn, False, False, 0)
        
        actions_box.pack_start(Gtk.Label(), True, True, 0)  # Spacer
        
        main_box.pack_start(actions_box, False, False, 0)
        
        post_frame.add(main_box)
        return post_frame
    
    # Event handlers
    def on_filter_changed(self, combo):
        self.current_filter = combo.get_active_text()
        # In a real implementation, this would filter posts
        self.status_label.set_text(f"Filter changed to {self.current_filter}")
    
    def on_notifications_clicked(self, button):
        dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO, Gtk.ButtonsType.OK,
                                 f"You have {self.unread_notifications} unread notifications")
        dialog.run()
        dialog.destroy()
    
    def on_trending_clicked(self, button):
        dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO, Gtk.ButtonsType.OK,
                                 "Top trending topics:\n#ReactNative\n#MobileDev\n#Tech\n#Startup\n#Innovation")
        dialog.run()
        dialog.destroy()
    
    def on_create_post(self, widget):
        content = self.create_post_entry.get_text().strip()
        if not content:
            return
        
        new_post = {
            'id': str(datetime.now().timestamp()),
            'user': {
                'id': 'current-user',
                'name': 'You',
                'username': '@you',
                'avatar': '👤',
                'verified': False
            },
            'content': content,
            'image': '',
            'location': '',
            'timestamp': 'now',
            'likes': 0,
            'comments': 0,
            'shares': 0,
            'is_liked': False,
            'is_bookmarked': False,
            'privacy': 'public'
        }
        
        self.posts.insert(0, new_post)
        self.all_posts.insert(0, new_post)
        self.create_post_entry.set_text("")
        self.update_posts_display()
    
    def on_like_clicked(self, button, post):
        # Toggle like
        post['is_liked'] = not post['is_liked']
        post['likes'] += 1 if post['is_liked'] else -1
        
        # Update in all_posts as well
        for all_post in self.all_posts:
            if all_post['id'] == post['id']:
                all_post['is_liked'] = post['is_liked']
                all_post['likes'] = post['likes']
                break
        
        self.update_posts_display()
    
    def on_comment_clicked(self, button, post):
        dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO, Gtk.ButtonsType.OK,
                                 f"Comments for: {post['user']['name']}\n\n{post['content']}")
        dialog.run()
        dialog.destroy()
    
    def on_share_clicked(self, button, post):
        dialog = Gtk.MessageDialog(self, 0, Gtk.MessageType.INFO, Gtk.ButtonsType.OK,
                                 f"Sharing: {post['title']}")
        dialog.run()
        dialog.destroy()
    
    def on_bookmark_clicked(self, button, post):
        # Toggle bookmark
        post['is_bookmarked'] = not post['is_bookmarked']
        
        # Update in all_posts as well
        for all_post in self.all_posts:
            if all_post['id'] == post['id']:
                all_post['is_bookmarked'] = post['is_bookmarked']
                break
        
        self.update_posts_display()

def main():
    app = SocialMediaFeed()
    app.connect("destroy", Gtk.main_quit)
    app.show_all()
    Gtk.main()

if __name__ == "__main__":
    main()
